src/Elastic.Apm/DiagnosticListeners/HttpDiagnosticListener.cs

```diff
@@ -117,8 +111,7 @@ private bool IsRequestFiltered(Uri requestUri)
 		{
 			switch (requestUri)
 			{
-				case Uri uri when uri == null:
-					return true;
+				case Uri uri when uri == null: return true;
 				case Uri uri when Agent.Config.ServerUrls.Any(n => n.IsBaseOf(uri)): //TODO: measure the perf of this!
```

Comment:
We have failing tests, and I think this line introduces a race condition. It can be fixed by passing an `AbstractConfig` to the constructor and using that instead of `Agent.Config.ServerUrls`. I already have the change locally on my machine.

Tests fail with: 
```
System.InvalidOperationException : Collection was modified; enumeration operation may not execute.
   at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source, Func`2 predicate)
   at Elastic.Apm.DiagnosticListeners.HttpDiagnosticListener.IsRequestFiltered(Uri requestUri) in /Users/..../Repos/apm-agent-dotnet/src/Elastic.Apm/DiagnosticListeners/HttpDiagnosticListener.cs:line 120

```

Reply:
Btw. it's quite unfortunate that those tests were green in the CI. I executed all tests locally on my machine and some were sporadically red, but only if no debugger was attached.

Reply:
By thinking further I'd say in the `AbstractAgentConfig` we need to introduce some locks for the initialization of the configsettings (btw. I think that's the ultimate solution to this problem). But that's unrelated to this change set, so I'd not focus on that one here.

Reply:
I ended up splitting `AbstractAgentConfig` into `AgentComponents` and `IConfigurationReader` to separate the two concerns.

`AgentComponents`  and `ApmAgent` both implement `IApmAgent` and can be used where something relies on one of the agent components. Both are internal so `Elastic.Apm.Agent` is still the only public facing way to setup and interact with the agent.

Reply:
👍